### PR TITLE
cross domain transaction recovery

### PR DESCRIPTION
**Pull request by bhavaniravichandran**
This test verifies cross domain transaction with Istio is successful and able to re-establish connection when
one domain is shutdown. domain in image using wdt is used to create 2 domains in different namespaces.
A servlet is deployed to the admin server of domain1. This servlet starts a transaction with
TMAfterTLogBeforeCommitExit transaction property set. Sends a message to a JMS queue and also inserts
data into oracleDB table. The coordinator (server in domain2) should exit before commit and the domain1
admin server should be able to establish connection with domain2 and the transaction should commit.

https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/2432/
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/2430/
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/2414/

### CODE REVIEW

File: integration-tests/src/test/resources/apps/cdtservlet/src/application/CdtTxServlet.java [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+      createTable(conn, tableName);
+
+      try {
+        tm.begin();
+        Transaction tx = (Transaction) tm.getTransaction();
+        Xid xid = tx.getXID();
+        System.out.println("xid=" + xid);
+        tx.setProperty("TMAfterTLogBeforeCommitExit", Boolean.TRUE);
+
+        ds = (javax.sql.DataSource) ctx2.lookup("jdbc/TestCdtDataSource");
```
Comment (ajsomogyi) [inline integration-tests/src/test/resources/apps/cdtservlet/src/application/CdtTxServlet.java:126]: You don't need to lookup the datasource again here.  Getting the connection under a global tx will cause the JDBC resource to be enlisted in the tx.

Comment (bhavaniravichandran) [reply integration-tests/src/test/resources/apps/cdtservlet/src/application/CdtTxServlet.java]: Fixed

File: integration-tests/src/test/resources/wdt-models/model-cdt-jdbc.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+          TestTableName: SQL SELECT 1
+        JDBCDataSourceParams:
+          GlobalTransactionsProtocol: TwoPhaseCommit
+          #GlobalTransactionsProtocol: LoggingLastResource
+          RowPrefetchSize: 200
+          JNDIName: jdbc/TestCdtDataSource
+        JDBCDriverParams:
+          URL: 'jdbc:oracle:thin:@//@@PROP:K8S_NODEPORT_HOST@@:@@PROP:DBPORT@@/devpdb.k8s'
+          PasswordEncrypted: 'Oradoc_db1'
+          #DriverName: com.mysql.cj.jdbc.MysqlXADataSource
```
Comment (anpanigr) [inline integration-tests/src/test/resources/wdt-models/model-cdt-jdbc.yaml:21]: remove commented code

File: integration-tests/src/test/resources/apps/cdtservlet/src/application/CdtTxServlet.java [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        System.out.println("CdTXServlet: Got PeerGoneException " + se);
+        out.println("CdTXServlet:Exception got - " + se.toString());
+        out.println("CdTXServlet: got PeerGoneException as expected");
+      }
+
+      out.println("Message Delivered in Distributed Transcation");
+
+      sleep(30);
+
+      String body = context.createConsumer(d).receiveBody(String.class);
```
Comment (anpanigr) [inline integration-tests/src/test/resources/apps/cdtservlet/src/application/CdtTxServlet.java:153]: Check the message  body , to see if the body contains the same message string.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItCrossDomainTransaction.java [hunk 4]
```diff
@@ -246,6 +303,41 @@ public void testCrossDomainTransaction() {
 
   }
 
+  /*
+   * This test verifies cross domain transaction is successful and able to re-establish connection when
+   * one domain is shutdown. domain in image using wdt is used to create 2 domains in different namespaces.
```
[the remark this answers is not in the document]
Comment (bhavaniravichandran) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItCrossDomainTransaction.java:308]: No, it is domain in image with wdt.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItCrossDomainTransaction.java [hunk 5]
```diff
@@ -246,6 +303,41 @@ public void testCrossDomainTransaction() {
...
+   * A servlet is deployed to the admin server of domain1. This servlet starts a transaction with
+   * TMAfterTLogBeforeCommitExit transaction property set. Sends a message to a JMS queue and also inserts
+   * data into oracleDB table. The coordinator (server in domain2) should exit before commit and the domain1
+   * admin server should be able to establish connection with domain2 and the transaction should commit.
```
[the remark this answers is not in the document]
Comment (bhavaniravichandran) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItCrossDomainTransaction.java:312]: Are you asking if I am restarting the server in the test? No. The server is shutdown and restarts - no user intervention necessary.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItCrossDomainTransaction.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItCrossDomainTransaction.java:308]: Just change the sentence which starts with a Lower case letter.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItCrossDomainTransaction.java [hunk 6]
```diff
@@ -246,6 +303,41 @@ public void testCrossDomainTransaction() {
...
+   *
+   */
+  @Test
+  @DisplayName("Check cross domain transaction works")
```
[the remark this answers is not in the document]
Comment (bhavaniravichandran) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItCrossDomainTransaction.java]: Done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItCrossDomainTransaction.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (bhavaniravichandran) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItCrossDomainTransaction.java]: Fixed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItCrossDomainTransaction.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (bhavaniravichandran) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItCrossDomainTransaction.java]: fixed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonTestUtils.java [hunk 7]
```diff
@@ -1740,21 +1740,21 @@ public static String createMiiImageAndVerify(String miiImageNameBase,
...
-                                            String appName,
+                                            List<String> wdtModelList,
+                                            List<String> appSrcDirList,
                                             String modelPropFile,
                                             String altModelDir,
                                             String domainHome) {
 
-    final List<String> wdtModelList = Collections.singletonList(MODEL_DIR + "/" + wdtModelFile);
-    final List<String> appSrcDirList = Collections.singletonList(appName);
+    //final List<String> wdtModelList = Collections.singletonList(MODEL_DIR + "/" + wdtModelFile);
```
[the remark this answers is not in the document]
Comment (bhavaniravichandran) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/utils/CommonTestUtils.java]: done

File: integration-tests/src/test/resources/wdt-models/model-cdt-jdbc.yaml (hunk 2 above)
Comment (bhavaniravichandran) [reply integration-tests/src/test/resources/wdt-models/model-cdt-jdbc.yaml]: done

File: integration-tests/src/test/resources/apps/cdtservlet/src/application/CdtTxServlet.java (hunk 3 above)
Comment (bhavaniravichandran) [reply integration-tests/src/test/resources/apps/cdtservlet/src/application/CdtTxServlet.java:154]: fixed.

**Review by ajsomogyi — APPROVED** [review]

**Review by anpanigr — APPROVED** [review]
LGTM

**Pull request merged**